### PR TITLE
Upgrade Chart.js engine from v1.0.1 to v2.3.0

### PR DESCRIPTION
**Pull request by DL6ER**
Changes proposed in this pull request:

- Upgrade Chart.js from version 1.0.1 (close to ancient) to 2.3.0 (current version).

As the syntax changed quite significantly, a major edit has been necessary. However, there is no conflict against the current development branch and an automatic merge is easily possible.

I think this upgrade is long overdue. I need the current version for my planed enhancements to the web interface. With the upgrade we can now use advanced chart options (like adding a second y axis, needed to make progress on issue #138).

Before this pull request:

old

And afterwards:

new

@pi-hole/dashboard

### CODE REVIEW

**Comment by Mcat12** [comment]
You beat me to it! NP though, check out [my half-way finished work](https://github.com/pi-hole/AdminLTE/tree/chartjs-upgrade) so that your fork can be even better.

**Comment by Mcat12** [comment]
Also, the query graph is too small on mobile: http://prntscr.com/d3xdnw

**Comment by DL6ER** [comment]
Whoops, I wasn't aware that someone else was also working on this topic. Anyhow, I added pointHitRadius and beginAtZero.

If you mean by "make the doughnut graphs a bit bigger" that we should make the margin (white space) above and below the doughnuts smaller, then I have to admit that I don't know how to do this (I've played like an hour with different settings, but nothing worked so far).

Nevertheless, I made everything a bit larger in terms of defined height, hopefully this will now work on small screens:

screenshot at 2016-11-06 22-40-31

**Comment by DL6ER** [comment]
On mobile. Not especially pretty, but at least it seems that everything fits on my Android device:
screenshot_20161106-224904

**Comment by Mcat12** [comment]
The graph could be a bit taller, by maybe x1.5 to x2

**Comment by DL6ER** [comment]
I made the graph larger and edited the title slightly as it wasn't clear "what" is actually shown there (requests per minute, hour, day ...). Eventually, I want to propose the make the data less averaged, but this is a different topic (although the code is already finished).

**Comment by Mcat12** [comment]
Check out [this](https://github.com/pi-hole/AdminLTE/tree/angularJS) branch for something like that.

**Comment by DL6ER** [comment]
I have the code for this other thing already done. I'll open another PR as soon as we finish this one.

File: js/pihole/index.js [hunk 1]
```diff
@@ -13,24 +13,6 @@ $(document).ready(function() {
 
     updateTopLists();
```
Comment (Mcat12) [inline js/pihole/index.js]: All of these functions, besides `updateSummaryData`, might be better after we create the charts, just to be sure the charts exist before we access them.

File: js/pihole/index.js [hunk 2]
```diff
@@ -110,11 +146,42 @@ function updateSummaryData(runOnce) {
...
+        // Add data for each hour that is available
         for (hour in data.ads_over_time) {
-            timeLineChart.addData([data.domains_over_time[hour], data.ads_over_time[hour]], hour + ":00");
+            // Add x-axis label
+            timeLineChart.data.labels.push(hour + ":00");
+            timeLineChart.data.datasets.forEach(function(dataset, index) {
+                // Add data to line with index 0 or 1
+                if(index == 0) dataset.data.push(data.domains_over_time[hour]);
+                if(index == 1) dataset.data.push(data.ads_over_time[hour]);
+            });
```
Comment (Mcat12) [inline js/pihole/index.js]: This `forEach` can be rewritten as this (I think, unless I'm missing something):
```
timeLineChart.data.datasets[0].data.push(data.domains_over_time[hour]);
timeLineChart.data.datasets[1].data.push(data.ads_over_time[hour]);
```

File: js/pihole/index.js [hunk 3]
```diff
@@ -110,11 +146,42 @@ function updateSummaryData(runOnce) {
...
+        });
+        // Build a single dataset with the data to be pushed
+        var dd = {data: v, backgroundColor: c};
+        // and push it at once
+        queryTypeChart.data.datasets.push(dd);
+        $('#query-types .overlay').remove();
+        queryTypeChart.update();
+        queryTypeChart.chart.config.options.animation.duration=2000;
+        queryTypeChart.chart.config.options.cutoutPercentage=30;
+        queryTypeChart.update();
```
Comment (Mcat12) [inline js/pihole/index.js]: `update` is called twice.

File: js/pihole/index.js [hunk 4]
```diff
@@ -131,35 +198,27 @@ function updateTopClientsChart() {
...
+        // Build a single dataset with the data to be pushed
+        var dd = {data: v, backgroundColor: c};
+        // and push it at once
+        forwardDestinationChart.data.datasets.push(dd);
         $('#forward-destinations .overlay').remove();
-        //$('#forward-destinations').append(forwardDestinationChart.generateLegend());
+        forwardDestinationChart.update();
+        forwardDestinationChart.chart.config.options.animation.duration=2000;
+        forwardDestinationChart.chart.config.options.cutoutPercentage=30;
+        forwardDestinationChart.update();
```
Comment (Mcat12) [inline js/pihole/index.js]: `update` is called twice.

File: js/pihole/index.js (hunk 1 above)
Comment (DL6ER) [inline js/pihole/index.js]: I doubt that the data would be delivered faster than the charts are generated, but nevertheless this seems a good idea.

File: js/pihole/index.js (hunk 2 above)
Comment (DL6ER) [inline js/pihole/index.js]: Indeed, I changed it.

File: js/pihole/index.js (hunk 4 above)
Comment (DL6ER) [inline js/pihole/index.js]: See above.

File: js/pihole/index.js [hunk 5]
```diff
@@ -1,17 +1,4 @@
...
-
-    updateQueriesOverTime();
-
-    updateQueryTypes();
-
-    updateTopClientsChart();
-
-    updateForwardDestinations();
-
-    updateTopLists();
```
Comment (DL6ER) [inline js/pihole/index.js]: Moved these function calls down.

File: js/pihole/index.js [hunk 6]
```diff
@@ -179,7 +185,6 @@ function updateQueryTypes() {
         queryTypeChart.data.datasets.push(dd);
         $('#query-types .overlay').remove();
         queryTypeChart.update();
-        queryTypeChart.chart.config.options.animation.duration=2000;
```
Comment (DL6ER) [inline js/pihole/index.js]: Moved this up (set it at creation time of the chart).

**Comment by Mcat12** [comment]
Approved

**Comment by PromoFaux** [comment]
Approved.

**Pull request merged**